packages/babel-helper-wrap-function/src/index.ts

```diff
@@ -179,19 +198,164 @@ function plainFunction(
+
+      // Unwrap the wrapper IIFE's environment so super and this and such still work.
+      (
+        path.get("body.body.0.argument.arguments.0") as NodePath
+      ).unwrapFunctionEnvironment();
+    } else {
+      let node;
+      let functionId = null;
+      const nodeParams = path.node.params;
+
+      if (path.isArrowFunctionExpression()) {
+        let path2;
+        if (process.env.BABEL_8_BREAKING) {
+          path2 = path.arrowFunctionToExpression({ noNewArrows });
+        } else {
+          // arrowFunctionToExpression returns undefined in @babel/traverse < 7.18.10
+          path2 = path.arrowFunctionToExpression({ noNewArrows }) ?? path;
+        }
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+        node = path2.node as
+          | t.FunctionDeclaration
+          | t.FunctionExpression
+          | t.CallExpression;
+      } else {
+        node = path.node as t.FunctionDeclaration | t.FunctionExpression;
+      }
+      const isDeclaration = path.isFunctionDeclaration();
+
+      let built = node;
+      if (!isCallExpression(node)) {
+        functionId = node.id;
+        built = callExpression(path.hub.addHelper(callAsync), [
+          functionExpression(null, node.params, node.body, node.generator),
+          identifier("this"),
+          identifier("arguments"),
+        ]);
+      }
+
+      const params: t.Identifier[] = [];
+      for (const param of nodeParams) {
+        if (isAssignmentPattern(param) || isRestElement(param)) {
+          break;
+        }
+        params.push(path.scope.generateUidIdentifier("x"));
+      }
+
+      let wrapper: t.Function = buildWrapper({
+        NAME: functionId,
+        CALL: built,
+        PARAMS: params,
+      });
+
+      if (!isDeclaration) {
+        wrapper = toExpression(wrapper);
+        nameFunction({
+          node: wrapper,
+          parent: (path as NodePath<t.FunctionExpression>).parent,
+          scope: path.scope,
+        });
+      }
+
+      if (
+        isDeclaration ||
+        wrapper.id ||
+        (!ignoreFunctionLength && params.length)
+      ) {
+        path.replaceWith(wrapper);
+        markCallWrapped(path);
+      } else {
+        // we can omit this wrapper as the conditions it protects for do not apply
+        path.replaceWith(
+          callExpression(path.hub.addHelper(callId as string), [
+            node as t.FunctionExpression,
+          ]),
+        );
+      }
+    }
+    return;
+  }
+
   if (path.isMethod()) {
-    classOrObjectMethod(path, callId);
+    classOrObjectMethod(path, callId as t.Expression);
   } else {
     plainFunction(
       path as NodePath<Exclude<t.Function, t.Method>>,
-      callId,
+      callId as t.Expression,
       noNewArrows,
       ignoreFunctionLength,
     );
   }
 }
+
+export function buildOnCallExpression(helperName: string) {
+  return {
+    CallExpression: {
+      exit(path: NodePath<t.CallExpression>, state: PluginPass) {
```

Comment:
https://github.com/babel/babel/pull/15992/files#diff-84a64030b885f05d8fef6726ef5a9c863a2761756c846ff5ca9278a058384151L1-L11
In the past, we always created a closure for the function to save, which was necessary for use cases like `babelHelpers.regeneratorRuntime().mark()`.
Now, since more and more users support generators natively, I chose not to create closures for all functions, but only save the results for use cases like `regeneratorRuntime`.
The generator function transformation runs after `wrapFunction`, and if we modify that plugin directly, we will have to deal with compatibility issues.
To be honest, after I finished it, I thought it was more complicated than I expected, but it's done and there's no harm in it.